### PR TITLE
move to the WIP Rust port of compiler-rt

### PR DESCRIPTION
**Pull request by japaric**
this replaces the in-tree compiler_builtins crate with the [out-of-tree WIP Rust port of compiler-rt](https://github.com/japaric/rustc-builtins).

_warning_ very lightly tested: make rustc-stage1 works with the Makefiles and with rustbuild but I haven't run make check yet.

I had to do some _interesting_ changes in the upstream "rustc-builtins" crate but I'll open a PR in that repository to discuss those.

One thing to still needs to be done is updating `bootstrap` to handle recursive submodules. I got the initialization of recursive submodules working but I haven't done the part that handles dirty/outdated recursive submodules.

cc @alexcrichton can we throw this into the try buildbots?
cc @brson 
cc @nagisa

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -17,3 +14,7 @@
 [submodule "src/liblibc"]
 	path = src/liblibc
 	url = https://github.com/rust-lang/libc.git
+[submodule "src/libcompiler_builtins"]
+	path = src/libcompiler_builtins
+	url = git://github.com/japaric/rustc-builtins.git
```
Comment (alexcrichton) [inline .gitmodules]: Before we land this I think we'll want to be sure to move this to rust-lang-nursery under the name "compiler-builtins"

**Comment by alexcrichton** [comment]
Sorry for being slow to take a look, want to rebase this and update with some recent changes and I'll throw it at the try bots?

**Comment by alexcrichton** [comment]
@japaric is this ready for the try bots? Sorry if I missed it!

Also one thought I just remembered I've been meaning to write down. We compile all C builtins with `-fvisibility=hidden` which I believe affects symbol visibility, especially when statically linking. I think, however, we're not doing the equivalent for the compiler-builtins crate written in Rust. This in turn might mean that the symbols are publicly exported from staticlibs, dylibs, etc, which ideally is something we'd prefer to avoid.

Could you take a look at this and ensure that we compile compiler-builtins Rust-defined symbols with `-fvisibility=hidden`

**Comment by brson** [comment]
+1

**Comment by nagisa** [comment]
NB: I’ll likely have to move this in-tree in order to get i128 intrinsics implemented, as we haven’t a way to stage out-of-tree dependencies AFAICT.

**Comment by japaric** [comment]
> Could you take a look at this and ensure that we compile compiler-builtins Rust-defined symbols with -fvisibility=hidden?

So, I checked and that's not the case. The Rust intrinsics appear as global/external symbols in dylibs (.so) whereas the C intrinsics don't. Seems like we'll have to do some compiler changes to change that behavior.

**Comment by alexcrichton** [comment]
Ok, closing due to inactivity for now. I've added an entry to https://github.com/rust-lang-nursery/compiler-builtins/issues/66 to track compiling the Rust versions with hidden visibility

**Comment by alexcrichton** [comment]
@japaric would you be interested in implementing such a patch to the compiler? If not I can take a look in the next few days

**Comment by japaric** [comment]
@alexcrichton All yours! (I've been meaning to look into this compiler change and how it relates to #37530 but I keep getting distracted with something else.)

**Comment by alexcrichton** [comment]
Ok, PR opened over at https://github.com/rust-lang/rust/pull/37714

**Comment by japaric** [comment]
Reopening now that #37714 landed. Rebased as well. Still testing locally but rustbuild is already past rustc-stage1.

**Pull request closed without merging**